File: haskell/qsort.hs
-- Quicksort workload — purely functional form: the classic list quicksort
-- (Haskell has no persistent array with efficient functional update in the
-- boot libraries, so lists are the idiomatic pure structure). Same MINSTD
-- fill and checksum as the other variants; element order before sorting is
-- irrelevant to the result.
{-# LANGUAGE BangPatterns #-}

module Main where

import Data.List (foldl')
import System.Exit (exitFailure)

lcg :: Int -> Int
lcg x = (x * 48271) `rem` 2147483647

fill :: Int -> [Int]
fill seed = go 65536 seed []
  where
    go :: Int -> Int -> [Int] -> [Int]
    go 0 _ acc = acc
    go !k !x acc = let x1 = lcg x in go (k - 1) x1 (x1 : acc)

qsort :: [Int] -> [Int]
qsort [] = []
qsort (p : xs) = qsort [x | x <- xs, x < p] ++ p : qsort [x | x <- xs, x >= p]

checksum :: [Int] -> Int
checksum = go (-1) 0
  where
    go _ !acc [] = acc
    go !prev !acc (v : vs)
      | prev > v = -1
      | otherwise = go v ((acc + v) `rem` 1000000007) vs

main :: IO ()
main = do
  let rounds = 400
      acc =
        foldl'
          (\ !a r -> (a + checksum (qsort (fill (42 + r)))) `rem` 1000000007)
          0
          (reverse [1 .. rounds])
  if acc /= 853505117
    then putStrLn ("FAIL: expected 853505117, got " ++ show acc) >> exitFailure
    else pure ()

File: rust/qsort.rs
// Functional-style quicksort workload, direct-mutation baseline: Vec<i64>
// filled from a MINSTD LCG, hand-written Lomuto partition (the same
// algorithm the other variants use — not sort_unstable), checksum of the
// sorted array, 100 rounds.

fn lcg(x: i64) -> i64 {
    (x * 48271) % 2147483647
}

fn fill(seed: i64) -> Vec<i64> {
    let mut v = Vec::with_capacity(65536);
    let mut x = seed;
    for _ in 0..65536 {
        x = lcg(x);
        v.push(x);
    }
    v
}

fn qsort(a: &mut [i64], lo: i64, hi: i64) {
    if lo >= hi {
        return;
    }
    let p = a[hi as usize];
    let mut i = lo;
    for j in lo..hi {
        if a[j as usize] < p {
            a.swap(i as usize, j as usize);
            i += 1;
        }
    }
    a.swap(i as usize, hi as usize);
    qsort(a, lo, i - 1);
    qsort(a, i + 1, hi);
}

fn checksum(a: &[i64]) -> i64 {
    let mut acc = 0i64;
    let mut prev = -1i64;
    for &v in a {
        if prev > v {
            return -1;
        }
        prev = v;
        acc = (acc + v) % 1000000007;
    }
    acc
}

fn main() {
    let rounds = 100i64;
    let mut acc = 0i64;
    for r in (1..=rounds).rev() {
        let mut a = fill(42 + r);
        qsort(&mut a, 0, 65535);
        acc = (acc + checksum(&a)) % 1000000007;
    }
    if acc != 276066679 {
        eprintln!("FAIL: expected 276066679, got {}", acc);
        std::process::abort();
    }
}
